src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -400,16 +419,153 @@ public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> s
             {
                 return new WhereSelectArrayIterator<TSource, TResult2>(_source, _predicate, CombineSelectors(_selector, selector));
             }
+        }
 
-            public override IEnumerable<TResult> Where(Func<TResult, bool> predicate)
+        internal class WhereSelectListIterator<TSource, TResult> : Iterator<TResult>
+        {
+            private readonly List<TSource> _source;
+            private readonly Func<TSource, bool> _predicate;
+            private readonly Func<TSource, TResult> _selector;
+            private List<TSource>.Enumerator _enumerator;
+
+            public WhereSelectListIterator(List<TSource> source, Func<TSource, bool> predicate, Func<TSource, TResult> selector)
             {
-                return new WhereEnumerableIterator<TResult>(this, predicate);
+                Debug.Assert(source != null);
+                Debug.Assert(predicate != null);
+                Debug.Assert(selector != null);
+                _source = source;
+                _predicate = predicate;
+                _selector = selector;
             }
 
-            public TResult[] ToArray()
+            public override Iterator<TResult> Clone()
+            {
+                return new WhereSelectListIterator<TSource, TResult>(_source, _predicate, _selector);
+            }
+
+            public override bool MoveNext()
+            {
+                switch (state)
+                {
+                    case 1:
+                        _enumerator = _source.GetEnumerator();
+                        state = 2;
+                        goto case 2;
+                    case 2:
+                        while (_enumerator.MoveNext())
+                        {
+                            TSource item = _enumerator.Current;
+                            if (_predicate(item))
+                            {
+                                current = _selector(item);
+                                return true;
+                            }
+                        }
+                        Dispose();
+                        break;
+                }
+                return false;
+            }
+
+            public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> selector)
+            {
+                return new WhereSelectListIterator<TSource, TResult2>(_source, _predicate, CombineSelectors(_selector, selector));
+            }
+        }
+
+        internal sealed class SelectEnumerableIterator<TSource, TResult> : Iterator<TResult>
```

Comment:
> It was though pretty quick to write by hand given WhereSelectEnumerableIterator as a starting point, and there's greater consistency between the different cases.

in general I'm less concerned in these cases about how hard it was to write, and more concerned with an increase in the amount of code that needs to be maintained.